### ISSUE TITLE
EOF error when using x509 certificate for login on HANA Cloud

### ISSUE THREAD

**Issue by duhaesbaert** [SAP/go-hdb#89]
I have created an app for connection/authenticating with HANA Cloud. The login through Basic Auth(user&pass) works correctly. but when I try to do the same using a certificate, the driver is returning me and EOF error, coming specifically from [line 171 on connection.go](https://github.com/SAP/go-hdb/blob/42f931ed683b26692b734142a773753d11d03854/driver/connection.go#L171):. This is the error message:
```
Connect to do HANA Cloud at 127.0.0.1:30041.
https://<USER>:<PASSWORD>@127.0.0.1:30041?TLSRootCAFile=%2Ftmp%2Fcert%2FEDUARDOCredentials.pem&TLSServerName=127.0.0.1%3A30041
hdb.driver 2022/04/13 17:10:35 connection.go:171: Connection write error local address 127.0.0.1:58188 remote address 127.0.0.1:30041: EOF
hdb.driver 2022/04/13 17:10:35 connection.go:171: Connection write error local address 127.0.0.1:58189 remote address 127.0.0.1:30041: EOF
hdb.driver 2022/04/13 17:10:36 connection.go:171: Connection write error local address 127.0.0.1:58193 remote address 127.0.0.1:30041: EOF
driver: bad connection
```
As can be seen by the logs, it is returning just like the message and the `err` var contains EOF. 

The function that I am executing is the following:
```go
sqlDB, err := sql.Open("hdb", uri.String())
if err != nil {
   return err
}

err = sqlDB.Ping() // <- this is where the error is returned from
if err != nil {
   return err
}
```

A couple of things that might be interesting to point out:

1. The error happens regardless of which certificate I send over, valid or invalid for that specific user. I have also tried with my DigiCert_Global_Root_CA as well, but same error;
2. As mentioned before, if you try to log in using only User&Password, the ping works fine;
3. The same does not happen through different applications, for example if you use [node hana-client](https://www.npmjs.com/package/@sap/hana-client);
4. I am running HANA Cloud using Docker.

**Comment by stfnmllr** [comment]
The error indicates that the connection is not established or broken. The error is returned by Ping() because sql.Open() itself is not creating a connection to the DB (please refer to https://pkg.go.dev/database/sql#Open).

"The error happens regardless of which certificate I send over, valid or invalid for that specific user"
Yes, the issue doesn't look like it would be certificate related. For testing you might omit this parameter.

The dataSourceName (DSN) used looks strange:
- there is no https schema - it's always hdb
- is 127.0.0.1:30041 really a SNI endpoint (looks like a standard non TLS port)?
- and if yes TLSServerName=127.0.0.1%3A30041 shouldn't include a port (nor any 'replacement' characters) but the server name SNI should select

As basic auth is working, which DSN did you use?

**Comment by duhaesbaert** [comment]
_"_The app is just for a validation of the functionality of using x509._"_ - The tests below are using hdb. Changed to https for another test and haven't changed back.

_"is 127.0.0.1:30041 really a SNI endpoint (looks like a standard non TLS port)?"_ - As I am running HANA Cloud on Kubernetes, I am executing a `port-forward` on port 30041 (hana-cloud-0 pod). This is why I am pointing to `127.0.0.1:30041`.

Now, as you mentioned TSLServerName should't include a port, when I try to use it without any ports, I have an error, and other than that, not exactly sure what I should be sending:
```
hdb://EDUARDO:<PASSWORD>@127.0.0.1:30041?TLSRootCAFile=%2Ftmp%2Fcert%2FEDUARDOCredentials.pem&TLSServerName=127.0.0.1
hdb.driver 2022/04/13 18:37:04 connection.go:171: Connection write error local address 127.0.0.1:59503 remote address 127.0.0.1:30041: x509: cannot validate certificate for 127.0.0.1 because it doesn't contain any IP SANs
hdb.driver 2022/04/13 18:37:04 connection.go:171: Connection write error local address 127.0.0.1:59504 remote address 127.0.0.1:30041: x509: cannot validate certificate for 127.0.0.1 because it doesn't contain any IP SANs
hdb.driver 2022/04/13 18:37:04 connection.go:171: Connection write error local address 127.0.0.1:59505 remote address 127.0.0.1:30041: x509: cannot validate certificate for 127.0.0.1 because it doesn't contain any IP SANs
driver: bad connection
```

As basic auth is working, which DSN did you use? - Similar to with a certificate, but then, I added `TLSInsecureSkipVerify`:
```
Connect to do HANA Cloud at 127.0.0.1:30041.
https://EDUARDO:<PASSWORD>@127.0.0.1:30041?TLSInsecureSkipVerify=
Connection established at 127.0.0.1:30041 for user EDUARDO.
```

**Comment by stfnmllr** [comment]
Ok - the TLSInsecureSkipVerify example proofs, that the endpoint is accepting a TLS connection (here again the schema is https - I do need to add a check to the driver :-)).

And maybe one additional remark - independently of
- connecting via tcp/ip directly or
- using an encrypted TLS connection (whether 'insecure' with TLSInsecureSkipVerify or checked with the certificate / server name 
the authentication 'against' the HANA database is always 'basic auth' (user / password) as go-hdb is only supporting this option.

Anyway, this time the error message 
x509: cannot validate certificate for 127.0.0.1 because it doesn't contain any IP SANs
clearly says, that the returned certificates are not valid for hostname 127.0.0.1 (which shouldn't be a surprise as this ip address stands for localhost).

As I do not know about the configuration of the HANA running in the docker container I can only assume that it's kind of the same SAP would use for their cloud instances. Here HAProxy is used and the TLSServerName is used by the SNI protocol to decide which 'backend' server ip address would be used. This can be double checked in the HAProxy.cfg residing in the docker container. And please see the link to the SAP help at https://github.com/SAP/go-hdb#hana-cloud-connection for the certificate creation. 

But as there is other options as well it's essential to have the information how the docker container and the HANA running as part of the docker container is set up.

And there is no internal magic done by go-hdb. You can use a connector object https://pkg.go.dev/github.com/SAP/go-hdb/driver#Connector and define the tls.Config to your needs (https://pkg.go.dev/github.com/SAP/go-hdb/driver#Connector.SetTLSConfig).
At https://pkg.go.dev/crypto/tls#Config under ServerName (which is set by TLSServerName) the comment says:
// ServerName is used to verify the hostname on the returned
// certificates unless InsecureSkipVerify is given. It is also included
// in the client's handshake to support virtual hosting unless it is
// an IP address.
Which does mean:
- in case of virtual hosting (SNI) the ServerName needs to be a 'hostname', not an ip address and
- the ServerName needs to match the hostname on the returned certificates

In summary: the given TLSServerName is invalid and we would need a correct hostname for a successful TLS connection or use TLSInsecureSkipVerify instead.

**Comment by duhaesbaert** [comment]
Got it. Thank you very much for the explanation!

I have looked through my environment, and the certificate that I am generating, and changed it to 'localhost', and it seems that this is the right one to use on this case. Upon running the tests, this is the result that I get now:
```
hdb://EDUARDO:pwd@127.0.0.1:30041?TLSRootCAFile=%2Ftmp%2Fcert%2FEDUARDOCredentials.pem&TLSServerName=localhost
Handling connection for 30041
hdb.driver 2022/04/13 20:42:59 connection.go:171: Connection write error local address 127.0.0.1:60839 remote address 127.0.0.1:30041: x509: certificate signed by unknown authority
Handling connection for 30041
hdb.driver 2022/04/13 20:42:59 connection.go:171: Connection write error local address 127.0.0.1:60840 remote address 127.0.0.1:30041: x509: certificate signed by unknown authority
Handling connection for 30041
hdb.driver 2022/04/13 20:42:59 connection.go:171: Connection write error local address 127.0.0.1:60841 remote address 127.0.0.1:30041: x509: certificate signed by unknown authority
driver: bad connection
```

So, it seems now that the failure is due to an issue with the certificate used.

**Comment by stfnmllr** [comment]
Hm, 
x509: certificate signed by unknown authority
looks like a self-signed certificate.

I am not sure what you are trying to accomplish with the certificate, but the TLSRootCAFile would be used to verify the server certificates and therefore you need to use the certificate mentioned in 
https://help.sap.com/docs/HANA_SERVICE_CF/cc53ad464a57404b8d453bbadbc81ceb/5bd9bcec690346a8b36df9161b1343c2.html?locale=en-US

also described in the README.md to this driver which I was mentioning above already:
...And please see the link to the SAP help at https://github.com/SAP/go-hdb#hana-cloud-connection for the certificate creation.

And for the TLSServerName you would need the HANA hostname (not localhost), something you should whether find as part of the documentation to the docker container you are using or as part of the HAProxy.cfg used 'within' the docker container in case this is setup like HANA cloud instances.

**Comment by duhaesbaert** [comment]
I am looking into the certificate setup.

One thing that I found interesting. I have created a HANA Cloud Trial, and I am using the same connection but poiting to that instance, and I am again seeing EOF, but now on the read operation:
```
Connect to do HANA Cloud at <instance>.hana.trial-us10.hanacloud.ondemand.com:443.
hdb://<user>:<pwd>@<instance>.hana.trial-us10.hanacloud.ondemand.com:443?TLSInsecureSkipVerify=
hdb.driver 2022/04/14 18:52:19 connection.go:151: Connection read error local address 10.14.148.43:56141 remote address 34.224.44.10:443: EOF
hdb.driver 2022/04/14 18:52:20 connection.go:151: Connection read error local address 10.14.148.43:56142 remote address 34.224.44.10:443: EOF
hdb.driver 2022/04/14 18:52:20 connection.go:151: Connection read error local address 10.14.148.43:56143 remote address 34.224.44.10:443: EOF
driver: bad connection
```

My function is basically this:
```go
func ConnectToDB(host, port, user, password, certPath string) error {
	fmt.Println("Connect to do HANA Cloud at " + host + ":" + port + ".")

	uri, err := buildURI(host, port, user, password, certPath)
	if err != nil {
		return err
	}

	//PRINT HBD QUERYSTRING FOR CONNECTION
	fmt.Println(uri.String())

	sqlDB, err := sql.Open("hdb", uri.String())
	if err != nil {
		return err
	}

	err = sqlDB.Ping()
	if err != nil {
		return err
	}

	fmt.Println("Connection established at " + host + ":" + port + " for user " + user + ".")
	return nil
}
```

**Comment by duhaesbaert** [comment]
Got it!
It explains what is going on. I believe I have made some confusion on the topic. My apologies on that.

I was looking to pass a x509 certificate for user authentication, but on this case the TLSRootCAFile is for Encrypted Communication, and not user authentication. 

Thank you for the help and explanation!

**Comment by stfnmllr** [comment]
With version 0.107.0 you can use X509 authentication as well. Please double check 
https://pkg.go.dev/github.com/SAP/go-hdb/driver#NewX509AuthConnector
https://pkg.go.dev/github.com/SAP/go-hdb/driver#NewX509AuthConnectorByFiles

**Comment by duhaesbaert** [comment]
Awesome, thank you for letting us know!
